bcm63xx: fix compilation with GCC11

neheb opened the pull request:
While C allows array decay, GCC11 is more strict about it regarding
function parameters. Fix the type to match board_get_mac_address.

Fixes:

error: '__builtin_memcpy' reading 6 bytes from a region of size 0

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb commented:
@rmilecki does this look right to you? I have no idea what a u8* cast does.

```
--- a/target/linux/bcm63xx/patches-5.4/532-MIPS-BCM63XX-add-inventel-Livebox-support.patch
+++ b/target/linux/bcm63xx/patches-5.4/532-MIPS-BCM63XX-add-inventel-Livebox-support.patch
@@ -58,7 +58,7 @@ Subject: [PATCH] MIPS: BCM63XX: add inventel Livebox support
  #endif /* __BOARD_COMMON_H */
 --- /dev/null
 +++ b/arch/mips/bcm63xx/boards/board_livebox.c
-@@ -0,0 +1,153 @@
+@@ -0,0 +1,154 @@
 +/*
 + * This file is subject to the terms and conditions of the GNU General Public
 + * License. See the file "COPYING" in the main directory of this archive
@@ -134,8 +134,9 @@ Subject: [PATCH] MIPS: BCM63XX: add inventel Livebox support
 +{
 +      u8 *p;
 +      int count;
++      u8 m[ETH_ALEN] = { 0xBE, 0xBF, 0xF3, 0x77 };
 +
-+      mac = (u8 *)0xBEBFF377;
++      memcpy(mac, m, ETH_ALEN);
 +
 +      p = mac + ETH_ALEN - 1;
 +      count = mac_addr_used;
```

Seems safer.

rmilecki commented on target/linux/bcm63xx/patches-5.10/532-MIPS-BCM63XX-add-inventel-Livebox-support.patch:
```diff
@@ -134,8 +134,9 @@ Subject: [PATCH] MIPS: BCM63XX: add inventel Livebox support
++	u8 m[ETH_ALEN] = { 0xBE, 0xBF, 0xF3, 0x77 };
 +
-+	memcpy(mac, (u8 *)0xBEBFF377, ETH_ALEN);
++	mac = m;
```
This looks wrong. You assign pointer to a local array and modify that array later. The whole point of `livebox_get_mac_address()` is to modify array pointed by a caller.

neheb replied:
A previous version of this patchset made mac a u8[ETH_ALEN] array, which is more correct anyway. But since it has nothing to do with the failure, I removed it.

rmilecki commented:
> `mac = (u8 *)0xBEBFF377;`

@neheb: where did you get that code from? I don't see it in original `532-MIPS-BCM63XX-add-inventel-Livebox-support.patch` or in code introduced by your `bcm63xx: fix compilation with GCC11`.

rmilecki commented:
```
u8 m[ETH_ALEN] = { 0xBE, 0xBF, 0xF3, 0x77 };

memcpy(mac, m, ETH_ALEN);
```
above is probably better / cleaner / safer than the original
```
memcpy(mac, (u8 *)0xBEBFF377, ETH_ALEN);
```

neheb commented:
> > `mac = (u8 *)0xBEBFF377;`
> 
> @neheb: where did you get that code from? I don't see it in original `532-MIPS-BCM63XX-add-inventel-Livebox-support.patch` or in code introduced by your `bcm63xx: fix compilation with GCC11`.

previous attempt at fixing this.

rmilecki commented on target/linux/bcm63xx/patches-5.10/532-MIPS-BCM63XX-add-inventel-Livebox-support.patch:
```diff
@@ -134,8 +134,9 @@ Subject: [PATCH] MIPS: BCM63XX: add inventel Livebox support
++	u8 m[ETH_ALEN] = { 0xBE, 0xBF, 0xF3, 0x77 };
 +
-+	memcpy(mac, (u8 *)0xBEBFF377, ETH_ALEN);
++	memcpy(mac, m, ETH_ALEN);;
```
A minor typo: `;;`

neheb replied:
Yeah haha.

rmilecki commented:
I don't have any `bcm63xx` device / Livebox to test it and guarantee it works. Looks good to me though.

@Noltari ?

neheb commented:
I played around with (u8 *)0xBEBFF377 in gotbolt. My conclusion is that if it worked, it was an accident. This is better.

Noltari commented:
> I don't have any `bcm63xx` device / Livebox to test it and guarantee it works. Looks good to me though.
> 
> @Noltari ?

I haven't got any Livebox to test it either. However, @danitool may be able to test that...

danitool commented:
> ```
> u8 m[ETH_ALEN] = { 0xBE, 0xBF, 0xF3, 0x77 };
> 
> memcpy(mac, m, ETH_ALEN);
> ```
This assigns the BE:BF:F3:77:00:00 address to the ethernet interface which isn't ok.

> 
> above is probably better / cleaner / safer than the original
> 
> ```
> memcpy(mac, (u8 *)0xBEBFF377, ETH_ALEN);
> ```
This one is ok. It picks the MAC address from a particular offset at the flash chip.

Therefore this commit you're trying to merge isn't ok.

rmilecki commented:
> > ```
> > memcpy(mac, (u8 *)0xBEBFF377, ETH_ALEN);
> > ```
> 
> This one is ok. It picks the MAC address from a particular offset at the flash chip.

Oh, I really didn't realize what is that code doing. It really should use some IO helper. Ideally we should replace it with NVMEM in DT.

neheb commented:
> > ```
> > u8 m[ETH_ALEN] = { 0xBE, 0xBF, 0xF3, 0x77 };
> > 
> > memcpy(mac, m, ETH_ALEN);
> > ```
> 
> 
> This assigns the BE:BF:F3:77:00:00 address to the ethernet interface which isn't ok.
> 
> 
> > above is probably better / cleaner / safer than the original
> > ```
> > memcpy(mac, (u8 *)0xBEBFF377, ETH_ALEN);
> > ```
> 
> 
> This one is ok. It picks the MAC address from a particular offset at the flash chip.
> 
> 
> Therefore this commit you're trying to merge isn't ok.

Got a solution for that GCC11 error?

neheb committed:
bcm63xx: remove memcpy from mac assignment

With GCC11, memcpy doesn't work here as it assumes a size of 0. Use
ioremap to avoid it.

Fixed parameter type to match board_get_mac_address.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb commented:
This compiles and removes temporary variable now.

rmilecki commented:
> It really should use some IO helper.

Try this:
```
void __iomem *mmio;

mmio = ioremap(0xbebff377, 0x8);
if (!mmio)
	return -EIO;
memcpy_fromio(mac, mmio, ETH_ALEN);
iounmap(mmio);
```

neheb commented:
Compiles.

rmilecki commented:
Looks good. Commit message should be adjusted in the `just use =` part.

neheb commented:
Done.

rmilecki commented:
Thanks. I can't test it but looks fine.

@danitool ?

neheb commented:
I was told by @danitool the address is wrong. Fixed it.

neheb commented:
Now this doesn't compile...

```
In file included from arch/mips/bcm63xx/boards/board_livebox.c:11:
In function 'memcpy',
    inlined from 'memcpy_fromio' at ./arch/mips/include/asm/io.h:586:2,
    inlined from 'livebox_get_mac_address' at arch/mips/bcm63xx/boards/board_livebox.c:81:2:
./include/linux/string.h:287:33: error: '__builtin_memcpy' may read 6 bytes from a region of size 0 [-Werror=stringop-overread]
  287 | #define __underlying_memcpy     __builtin_memcpy
      |                                 ^
./include/linux/string.h:403:16: note: in expansion of macro '__underlying_memcpy'
  403 |         return __underlying_memcpy(p, q, size);
      |                ^~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
make[8]: *** [scripts/Makefile.build:262: arch/mips/bcm63xx/boards/board_livebox.o] Error 1

```

neheb commented:
Works when adding volatile.

```
volatile void __iomem *mmio;
```
fails.

```
void __iomem *volatile mmio;
```

succeeds...

danitool commented:
Tested and working ok.

Thank you.

aparcar commented:
Merged, thanks!
